src/main/java/org/sonarsource/sonarlint/ls/AnalysisManager.java

```diff
@@ -0,0 +1,452 @@
+
+  private final SonarLintTelemetry telemetry;
+  private final LanguageClientLogOutput clientLogOutput;
+  private final Collection<URL> standaloneAnalyzers;
+  private final WorkspaceFoldersManager workspaceFoldersManager;
+  private final SettingsManager settingsManager;
+  private final ProjectBindingManager bindingManager;
+  private final EventWatcher watcher;
+  private final int timerMs;
+  @CheckForNull
+  private Path typeScriptPath;
+  private StandaloneSonarLintEngine standaloneEngine;
+
+  private ExecutorService analysisExecutor;
+
+  public AnalysisManager(Collection<URL> standaloneAnalyzers, LanguageClientLogOutput clientLogOutput, SonarLintExtendedLanguageClient client,
+    SonarLintTelemetry telemetry,
+    WorkspaceFoldersManager workspaceFoldersManager, SettingsManager settingsManager, ProjectBindingManager bindingManager) {
+    this.standaloneAnalyzers = standaloneAnalyzers;
+    this.clientLogOutput = clientLogOutput;
+    this.client = client;
+    this.telemetry = telemetry;
+    this.workspaceFoldersManager = workspaceFoldersManager;
+    this.settingsManager = settingsManager;
+    this.bindingManager = bindingManager;
+    this.timerMs = DEFAULT_TIMER_MS;
+    this.analysisExecutor = Executors.newSingleThreadExecutor(Utils.threadFactory("SonarLint analysis", false));
+    this.watcher = new EventWatcher();
+  }
+
+  synchronized StandaloneSonarLintEngine getOrCreateStandaloneEngine() {
+    if (standaloneEngine == null) {
+      standaloneEngine = createStandaloneEngine();
+    }
+    return standaloneEngine;
+  }
+
+  private StandaloneSonarLintEngine createStandaloneEngine() {
+    LOG.debug("Starting standalone SonarLint engine...");
+    LOG.debug("Using {} analyzers", standaloneAnalyzers.size());
+
+    try {
+      Map<String, String> extraProperties = new HashMap<>();
+      if (typeScriptPath != null) {
+        extraProperties.put(TYPESCRIPT_PATH_PROP, typeScriptPath.toString());
+      }
+      StandaloneGlobalConfiguration configuration = StandaloneGlobalConfiguration.builder()
+        .setExtraProperties(extraProperties)
+        .addPlugins(standaloneAnalyzers.toArray(new URL[0]))
+        .setLogOutput(clientLogOutput)
+        .build();
+
+      StandaloneSonarLintEngine engine = new StandaloneSonarLintEngineImpl(configuration);
+      LOG.debug("Standalone SonarLint engine started");
+      return engine;
+    } catch (Exception e) {
+      LOG.error("Error starting standalone SonarLint engine", e);
+      throw new IllegalStateException(e);
+    }
+  }
+
+  public void didOpen(URI fileUri, String languageId, String fileContent) {
+    languageIdPerFileURI.put(fileUri, languageId);
+    fileContentPerFileURI.put(fileUri, fileContent);
+    analyzeAsync(fileUri, true);
+  }
+
+  public void didChange(URI fileUri, String fileContent) {
+    fileContentPerFileURI.put(fileUri, fileContent);
+    eventMap.put(fileUri, System.currentTimeMillis());
```

Comment:
Would it make sense to use `System2.now()`?

Reply:
Not if we don't plan to control the time in a test. And we can use the official Clock class if needed.